### ISSUE TITLE
training elmo from scratch inside another model

### ISSUE THREAD

**Issue by rajarsheem** [allenai/allennlp#2274]
Hi,
I am trying to use the elmo module inside my own model and want the embeddings to get jointly trained from scratch along with the parameters of the main model ABC without using the pre-initialized vectors (and then fine-tuning it). 
What I have done below is that I have loaded elmo module with pre-trained vectors as usual followed by re-initializing the weights inside the elmo module with the normal initializer.

Is this the recommended way to accomplish this? If not, can you please suggest a better way?

```python
class ABC(nn.Module):
    def __init__(self):
        super(Ranker, self).__init__()
        self.elmo = Elmo(options_file, weight_file, dropout=0.1, num_output_representations=1, requires_grad=True)
        self.reset_parameters()
    
    def forward(self, inp):
        return None
    
    def reset_parameters(self):
        for param in self.parameters():
            if not param.requires_grad: # embedding                
                continue
            if param.dim() < 2: # bias
                nn.init.constant_(param, 0)
            else:
                nn.init.normal_(param)
```

**Comment by matt-gardner** [comment]
@rajarsheem, yes, this could work, except `normal` isn't the best initialization for these things (e.g., we use [`block_orthogonal` and other options](https://github.com/allenai/allennlp/blob/259ce3221b5ed35c94ba4eaae070e22947f23f7b/allennlp/modules/lstm_cell_with_projection.py#L83-L91)).  You'd probably be better off if we just make the `weights_file` an optional argument to `Elmo`.  PRs welcome for that.

**Comment by matt-gardner** [comment]
You may also be able to dig in to the internal fields and call that `reset_parameters` function yourself, instead of the `normal` init that you're doing.

**Comment by matt-peters** [comment]
You could do that, but you'll probably get subpar results.  The Elmo class returns a linear combination of the layers, but if you are training from scratch (or fine tuning) just use the top layer.  You can accomplish this by passing `scalar_mix_parameters=[0.0, 0.0, 20.0]` when constructing `Elmo` class, see: https://github.com/allenai/allennlp/blob/master/allennlp/modules/elmo.py#L74  (we take softmax of parameters, so this effectively selects the top layer).

Or you can use the lower level components (`ElmoBilm` and `_ElmoCharacterEncoder`) directly.

**Comment by rajarsheem** [comment]
Hey, I made just two changes: When ```weight_file``` is not ```None```, only then load weights for ```_ElmoCharacterEncoder``` and ```_ElmoBiLm``` . This passes when I provide ```weight_file``` as None and ```options_file``` containing the default parameters that have been provided. 

Assuming that the initializations are fine, this should work right?

**Comment by rajarsheem** [comment]
@matt-peters , Can you please explain a bit why you are suggesting to use only the top layer when one is trying to initialize and train from scratch?

**Comment by rajarsheem** [comment]
Hi, @matt-gardner , I am trying to make ```weights_file``` optional. 
I noticed that there is no code for initialization of the weights for ```_ElmoCharacterEncoder``` -- only for loading is given.

How do you suggest I initialize them?

**Comment by rajarsheem** [comment]
@matt-peters Can you please explain a bit why you are suggesting to use only the top layer when one wants to initialize and train from scratch?

**Comment by rajarsheem** [comment]
As a temporary solution, I have sort of re-initialized post loading the pre-trained weights:

```python
    def _reset_elmo(self):
        filters = [[1, 32], [2, 32], [3, 64], [4, 128], [5, 256], [6, 512], [7, 1024]]
        highway_dim = n_filters = sum(f[1] for f in filters)

        nn.init.uniform_(self.elmo._elmo_lstm._token_embedder._char_embedding_weights, -1.0, 1.0)
        for conv in self.elmo._elmo_lstm._token_embedder._convolutions:
            nn.init.uniform_(conv.weight, -0.05, 0.5)
            nn.init.zeros_(conv.bias)

        for layer in self.elmo._elmo_lstm._token_embedder._highways._layers:
            nn.init.normal_(layer.weight, mean=0.0, std=np.sqrt(1.0/highway_dim))
            nn.init.zeros_(layer.bias)

        nn.init.normal_(self.elmo._elmo_lstm._token_embedder._projection.weight, mean=0.0, std=np.sqrt(1.0/n_filters))
        nn.init.zeros_(self.elmo._elmo_lstm._token_embedder._projection.bias)

        for layer in self.elmo._elmo_lstm._elmo_lstm.forward_layers:
            layer.reset_parameters()

        for layer in self.elmo._elmo_lstm._elmo_lstm.backward_layers:
            layer.reset_parameters()
```

I have kept the initializers same as what used in [bilm-tf](https://github.com/allenai/bilm-tf/blob/master/bilm/model.py)

**Comment by matt-gardner** [comment]
That temporary solution looks fine to me, assuming you got all of the initializers right (I didn't check that).  If you're up to it, we'd also be happy to take a PR that adds in the initialization for whatever's missing it, and makes the weights file optional.

**Comment by matt-peters** [comment]
+1 on the PR, this is a great addition.

**Comment by rajarsheem** [comment]
Yeah, I will probably make the changes for the PR once I get free
@matt-peters , still curious about why you suggested to use only the top layer!
